File: array/next-permutation.py
from typing import List

class Solution:
    def findBreakingPoint(self, nums: List[int]) -> int:
        for i in range(len(nums)-1, 0, -1):
            if nums[i] > nums[i-1]:
                return i-1
        return -1

    def nextPermutation(self, nums: List[int]) -> None:
        if len(nums) <= 1:
            return

        breaking_point = self.findBreakingPoint(nums)

        if breaking_point == -1:
            nums.reverse()
            return

        for i in range(len(nums)-1, breaking_point, -1):
            if nums[i] > nums[breaking_point]:
                nums[i], nums[breaking_point] = nums[breaking_point], nums[i]
                break

        nums[breaking_point+1:] = reversed(nums[breaking_point+1:])

File: 31-next-permutation/next-permutation.go
func reverse(nums []int) {
    l := 0
    r := len(nums) - 1
    for l < r {
        nums[l], nums[r] = nums[r], nums[l]
        l++
        r--
    }
}

func nextPermutation(nums []int) {
    n := len(nums)
    ind := -1
    for i := n - 1; i > 0; i-- {
        if nums[i-1] < nums[i] {
            ind = i
            break
        }
    }
    if ind == -1 {
        reverse(nums)
        return
    }
    for i := n - 1; i >= ind; i-- {
        if nums[i] > nums[ind-1] {
            nums[i], nums[ind-1] = nums[ind-1], nums[i]
            break
        }
    }
    reverse(nums[ind:])
}